File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/ui/views/ScreenSaverView.java
/*
 * <!--
 *   ~ Copyright (c) 2017. ThanksMister LLC
 *   ~
 *   ~ Licensed under the Apache License, Version 2.0 (the "License");
 *   ~ you may not use this file except in compliance with the License. 
 *   ~ You may obtain a copy of the License at
 *   ~
 *   ~ http://www.apache.org/licenses/LICENSE-2.0
 *   ~
 *   ~ Unless required by applicable law or agreed to in writing, software distributed 
 *   ~ under the License is distributed on an "AS IS" BASIS, 
 *   ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *   ~ See the License for the specific language governing permissions and 
 *   ~ limitations under the License.
 *   -->
 */

package com.thanksmister.iot.mqtt.alarmpanel.ui.views;

import android.content.Context;
import android.os.Handler;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.thanksmister.iot.mqtt.alarmpanel.R;
import com.thanksmister.iot.mqtt.alarmpanel.network.InstagramApi;
import com.thanksmister.iot.mqtt.alarmpanel.network.fetchers.InstagramFetcher;
import com.thanksmister.iot.mqtt.alarmpanel.network.model.InstagramItem;
import com.thanksmister.iot.mqtt.alarmpanel.network.model.InstagramResponse;
import com.thanksmister.iot.mqtt.alarmpanel.tasks.DarkSkyTask;
import com.thanksmister.iot.mqtt.alarmpanel.tasks.InstagramTask;

import java.text.DateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Random;

import butterknife.Bind;
import butterknife.ButterKnife;
import retrofit2.Response;
import timber.log.Timber;

public class ScreenSaverView extends RelativeLayout {
    
    @Bind(R.id.screenSaverImage)
    ImageView screenSaverImage;

    @Bind(R.id.screenSaverClock)
    TextView screenSaverClock;
    
    private InstagramTask task;
    private String userName;
    private boolean fitToScreen;
    private Handler rotationHandler;
    private Handler timeHandler;
    private Picasso picasso;
    private List<InstagramItem> itemList;
    private Context context;
    private String imageUrl;
    private long rotationInterval;

    public ScreenSaverView(Context context) {
        super(context);
        this.context = context;
    }

    public ScreenSaverView(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);
    }
    
    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if(task != null) {
            task.cancel(true);
            task = null;
        }

        if(picasso != null) {
            picasso.invalidate(imageUrl);
            picasso.cancelRequest(screenSaverImage);
            picasso = null;
        }

        if(rotationHandler != null) {
            rotationHandler.removeCallbacks(delayRotationRunnable);
        }
        
        if(timeHandler != null) {
            timeHandler.removeCallbacks(timeRunnable);
        }
    }
    
    public void setScreenSaver(Context context, boolean useImageScreenSaver, String userName, boolean fitToScreen, 
                               int rotationIntervalMinutes) {
        this.context = context;
        this.userName = userName;
        this.fitToScreen = fitToScreen;
        this.rotationInterval = rotationIntervalMinutes*1000; // convert to milliseconds
       
        if(useImageScreenSaver && !TextUtils.isEmpty(userName) ) {
            screenSaverImage.setVisibility(View.VISIBLE);
            screenSaverClock.setVisibility(View.GONE);
            if(timeHandler != null) {
                timeHandler.removeCallbacks(timeRunnable);
            }
            startScreenSavor();
        } else { // use clock
            screenSaverImage.setVisibility(View.GONE);
            screenSaverClock.setVisibility(View.VISIBLE);
            timeHandler = new Handler();
            timeHandler.postDelayed(timeRunnable, 10);
        }
    }
    
    private void startScreenSavor() {
        if(itemList == null || itemList.isEmpty()) {
            fetchMediaData();
        } else {
            startImageRotation();
        }
    }

    private Runnable delayRotationRunnable = new Runnable() {
        @Override
        public void run() {
            rotationHandler.removeCallbacks(delayRotationRunnable);
            startImageRotation();
        }
    };

    private Runnable timeRunnable = new Runnable() {
        @Override
        public void run() {
            String currentTimeString = DateFormat.getTimeInstance(DateFormat.DEFAULT, Locale.getDefault()).format(new Date());
            screenSaverClock.setText(currentTimeString);
            if(timeHandler != null) {
                timeHandler.postDelayed(timeRunnable, 1000);
            }
        }
    };

    private void startImageRotation() {
        if(picasso == null) {
            picasso = Picasso.with(context);
        }
        if(itemList != null && !itemList.isEmpty()) {
            final int min = 0;
            final int max = itemList.size() - 1;
            final int random = new Random().nextInt((max - min) + 1) + min;
            InstagramItem instagramItem = itemList.get(random);
            imageUrl = instagramItem.getImages().getStandardResolution().getUrl();
            if(fitToScreen) {
                picasso.load(imageUrl)
                        .placeholder(R.color.black)
                        .resize(screenSaverImage.getWidth(), screenSaverImage.getHeight())
                        .centerCrop()
                        .error(R.color.black)
                        .into(screenSaverImage);
            } else {
                picasso.load(imageUrl)
                        .placeholder(R.color.black)
                        .error(R.color.black)
                        .into(screenSaverImage);
            }
            if(rotationHandler == null) {
                rotationHandler = new Handler();
            }
            rotationHandler.postDelayed(delayRotationRunnable, rotationInterval);
        }
    }
    
    private void fetchMediaData() {
        if(task == null || task.isCancelled()) {
            final InstagramApi api = new InstagramApi();
            final InstagramFetcher fetcher = new InstagramFetcher(api);
            task = new InstagramTask(fetcher);
            task.setOnExceptionListener(new DarkSkyTask.OnExceptionListener() {
                public void onException(Exception exception) {
                    Timber.e("Instagram Exception: " + exception.getMessage());
                }
            });
            task.setOnCompleteListener(new InstagramTask.OnCompleteListener<Response<InstagramResponse>>() {
                public void onComplete(Response<InstagramResponse> response) {
                    Timber.d("Response: " + response);
                    Timber.d("Response: " + response.code());
                    InstagramResponse instagramResponse = response.body();
                    Timber.d("InstagramResponse: " + instagramResponse);
                    if (instagramResponse != null) {
                        itemList = instagramResponse.getItems();
                        Timber.d("itemList: " + itemList.size());
                        startImageRotation();
                    }
                }
            });
            task.execute(userName);
        }
    }
}

File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/ui/views/ScreenSaverView.kt
/*
 * Copyright (c) 2018 ThanksMister LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thanksmister.iot.mqtt.alarmpanel.ui.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.net.http.SslError
import android.os.Build
import android.os.Handler
import android.text.TextUtils
import android.text.format.DateUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import android.webkit.*
import android.widget.RelativeLayout
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.core.content.res.ResourcesCompat
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.NetworkPolicy
import com.squareup.picasso.Picasso
import com.thanksmister.iot.mqtt.alarmpanel.BaseActivity
import com.thanksmister.iot.mqtt.alarmpanel.R
import com.thanksmister.iot.mqtt.alarmpanel.network.ImageOptions
import com.thanksmister.iot.mqtt.alarmpanel.persistence.Configuration
import com.thanksmister.iot.mqtt.alarmpanel.persistence.Weather
import com.thanksmister.iot.mqtt.alarmpanel.persistence.WeatherDao
import com.thanksmister.iot.mqtt.alarmpanel.tasks.ImageTask
import com.thanksmister.iot.mqtt.alarmpanel.utils.StringUtils
import com.thanksmister.iot.mqtt.alarmpanel.utils.WeatherUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.adapter_forcast_card.view.*
import kotlinx.android.synthetic.main.dialog_screen_saver.view.*
import kotlinx.android.synthetic.main.dialog_screen_saver.view.temperatureText
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit

class ScreenSaverView : RelativeLayout {

    private val disposable = CompositeDisposable()
    private var task: ImageTask? = null
    private var rotationHandler: Handler? = null
    private var timeHandler: Handler? = null
    private var picasso: Picasso? = null
    private var imageUrl: String? = null
    private var rotationInterval: Long = 0
    private var saverContext: Context? = null
    private var weatherSource: WeatherDao? = null
    private var useUnsplashScreenSaver: Boolean = false
    private var useWebScreenSaver: Boolean = false
    private var useClockScreenSaver: Boolean = false
    private var hasWeather: Boolean = false
    private var useImperial: Boolean = false
    private var parentWidth: Int = 0
    private var parentHeight: Int = 0
    private val calendar: Calendar = Calendar.getInstance()
    private var delayRotate = true
    private var webUrl = Configuration.WEB_SCREEN_SAVER
    private var certPermissionsShown = false

    private val delayRotationRunnable = object : Runnable {
        override fun run() {
            startImageScreenSavor()
            rotationHandler?.postDelayed(this, TimeUnit.SECONDS.toMillis(rotationInterval))
        }
    }

    private val timeRunnable = object : Runnable {
        override fun run() {
            val date = Date()
            calendar.time = date
            val currentTimeString = DateUtils.formatDateTime(context, date.time, DateUtils.FORMAT_SHOW_TIME)
            screenSaverClock.text = currentTimeString
            screenSaverClockSmall.text = currentTimeString

            // use this only with the clock feature
            if (!useUnsplashScreenSaver && !useWebScreenSaver && !delayRotate) {
                val width = screenSaverClockLayout.width
                val height = screenSaverClockLayout.height
                parentWidth = screenSaverView.width
                parentHeight = screenSaverView.height
                try {
                    if (width > 0 && height > 0 && parentWidth > 0 && parentHeight > 0) {
                        if (parentHeight - width > 0) {
                            val newX = Random().nextInt(parentWidth - width)
                            screenSaverClockLayout.x = newX.toFloat()
                        }
                        if (parentHeight - height > 0) {
                            val newY = Random().nextInt(parentHeight - height)
                            screenSaverClockLayout.y = newY.toFloat()
                        }
                    }
                } catch (e: IllegalArgumentException) {
                    Timber.e(e.message)
                }
            }

            val offset = 60L - calendar.get(Calendar.SECOND)
            timeHandler?.postDelayed(this, TimeUnit.SECONDS.toMillis(offset))
            delayRotate = false
        }
    }

    constructor(context: Context) : super(context) {
        saverContext = context
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        saverContext = context
    }

    fun init(useUnsplashScreenSaver: Boolean = false,
             useClockScreenSaver: Boolean = false,
             hasWeather: Boolean = false,
             useImperial: Boolean = false,
             hasWebScreenSaver: Boolean = false,
             imageOptions: ImageOptions,
             webUrl: String = "",
             weatherDao: WeatherDao) {

        this.weatherSource = weatherDao
        this.rotationInterval = imageOptions.imageRotation.toLong()
        this.hasWeather = hasWeather
        this.useImperial = useImperial
        this.useUnsplashScreenSaver = useUnsplashScreenSaver
        this.useWebScreenSaver = hasWebScreenSaver
        this.useClockScreenSaver = useClockScreenSaver
        if (!TextUtils.isEmpty(webUrl)) {
            this.webUrl = webUrl
        }
        setScreenSaverView()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (task != null) {
            task!!.cancel(true)
            task = null
        }

        if (picasso != null) {
            picasso!!.invalidate(imageUrl!!)
            picasso!!.cancelRequest(screenSaverImage)
            picasso = null
        }

        if (rotationHandler != null) {
            rotationHandler!!.removeCallbacks(delayRotationRunnable)
        }

        if (timeHandler != null) {
            timeHandler!!.removeCallbacks(timeRunnable)
        }

        disposable.clear()
    }

    private fun setScreenSaverView() {
        val initialRegular = screenSaverClock.textSize
        val initialSmall = screenSaverClockSmall.textSize

        if (!hasWeather) {
            screenSaverClock.setTextSize(TypedValue.COMPLEX_UNIT_PX, initialRegular + 100)
            screenSaverClockSmall.setTextSize(TypedValue.COMPLEX_UNIT_PX, initialSmall + 60)
        }

        if (useUnsplashScreenSaver) {
            screenSaverImageLayout.visibility = View.VISIBLE
            screenSaverWebViewLayout.visibility = View.GONE
            screenSaverWeatherSmallLayout.visibility = View.GONE
            screenSaverClockSmall.visibility = View.GONE
            if (hasWeather && useClockScreenSaver) {
                setWeatherDataOnView()
                screenSaverWeatherSmallLayout.visibility = View.VISIBLE
                screenSaverClockSmall.visibility = View.VISIBLE
                timeHandler = Handler()
                timeHandler?.postDelayed(timeRunnable, 10)
            }
            rotationHandler = Handler()
            rotationHandler?.postDelayed(delayRotationRunnable, 10)
        } else if (useWebScreenSaver && webUrl.isNotEmpty()) {
            screenSaverWebViewLayout.visibility = View.VISIBLE
            screenSaverImageLayout.visibility = View.GONE
            screenSaverClockLayout.visibility = View.GONE
            if (hasWeather && useClockScreenSaver) {
                setWeatherDataOnView()
                screenSaverWeatherSmallLayout.visibility = View.VISIBLE
                screenSaverClockSmall.visibility = View.VISIBLE
                timeHandler = Handler()
                timeHandler?.postDelayed(timeRunnable, 10)
            }
            startWebScreenSaver(webUrl)
        } else if (useClockScreenSaver) { // use clock
            screenSaverWebViewLayout.visibility = View.GONE
            screenSaverImageLayout.visibility = View.GONE
            screenSaverClockLayout.visibility = View.VISIBLE
            if (!hasWeather) {
                screenSaverWeatherLayout.visibility = View.GONE
            } else {
                setWeatherDataOnView()
                screenSaverWeatherLayout.visibility = View.VISIBLE
            }
            timeHandler = Handler()
            timeHandler?.postDelayed(timeRunnable, 10)
        }
    }

    @SuppressLint("CheckResult")
    private fun setWeatherDataOnView() {
        weatherSource?.let {
            disposable.add(it.getItems()
                    .filter { items -> items.isNotEmpty() }
                    .map { items -> items[0] }
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnError { Weather() }
                    .subscribe { item ->
                        setDisplayData(item)
                    })
        }
    }

    private fun setDisplayData(weather: Weather) {
        val displayUnits = if (useImperial) saverContext!!.getString(R.string.text_f) else saverContext!!.getString(R.string.text_c)
        val precipitation = weather.precipitation.orEmpty()
        if (useUnsplashScreenSaver) {
            temperatureTextSmall.text = saverContext?.getString(R.string.text_temperature, weather.temperature.toString(), displayUnits)
            temperatureTextSmall.text = saverContext?.getString(R.string.text_temperature, weather.temperature.toString(), displayUnits)
            if (StringUtils.isDouble(precipitation) && shouldTakeUmbrellaToday(StringUtils.stringToDouble(precipitation))) {
                conditionImageSmall.setImageDrawable(ResourcesCompat.getDrawable(resources, R.drawable.ic_rain_umbrella, (saverContext!! as BaseActivity).theme))
            } else {
                conditionImageSmall.setImageDrawable(ResourcesCompat.getDrawable(resources, WeatherUtils.getIconForWeatherCondition(weather.condition), (saverContext!! as BaseActivity).theme))
            }
        } else {
            temperatureText.text = saverContext?.getString(R.string.text_temperature, weather.temperature.toString(), displayUnits)
            if (StringUtils.isDouble(precipitation) && shouldTakeUmbrellaToday(StringUtils.stringToDouble(precipitation))) {
                conditionImage.setImageDrawable(ResourcesCompat.getDrawable(resources, R.drawable.ic_rain_umbrella, (saverContext!! as BaseActivity).theme))
            } else {
                conditionImage.setImageDrawable(ResourcesCompat.getDrawable(resources, WeatherUtils.getIconForWeatherCondition(weather.condition), (saverContext!! as BaseActivity).theme))
            }
        }
    }

    private fun shouldTakeUmbrellaToday(precipitation: Double?): Boolean {
        precipitation?.let {
            return precipitation > PRECIP_AMOUNT
        }
        return false
    }

    private fun closeView() {
        this.callOnClick()
    }

    private fun startWebScreenSaver(url: String) {
        Timber.d("startWebScreenSaver $url")
        loadWebPage(url)
    }

    private fun startImageScreenSavor() {
        Picasso.get()
                .load(String.format(UNSPLASH_IT_URL, screenSaverView.width, screenSaverView.height))
                .memoryPolicy(MemoryPolicy.NO_CACHE)
                .networkPolicy(NetworkPolicy.NO_CACHE)
                .into(screenSaverImage)
    }

    private fun loadWebPage(url: String) {
        Timber.d("loadWebPage url ${url}")
        configureWebSettings("")
        clearCache()
        screenSaverWebView?.webChromeClient = object : WebChromeClient() {
            override fun onJsAlert(view: WebView, url: String, message: String, result: JsResult): Boolean {
                AlertDialog.Builder(view.context, R.style.CustomAlertDialog)
                        .setTitle(context.getString(R.string.dialog_title_ssl_error))
                        .setMessage(message)
                        .setPositiveButton(android.R.string.ok, null)
                        .show()
                return true
            }
        }
        screenSaverWebView?.setOnTouchListener(object : View.OnTouchListener {
            override fun onTouch(v: View?, event: MotionEvent?): Boolean {
                v?.performClick()
                closeView()
                return false
            }
        })
        screenSaverWebView?.webViewClient = object : WebViewClient() {
            //If you will not use this method url links are open in new browser not in webview
            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                view.loadUrl(url)
                return true
            }

            override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
                Toast.makeText(context, description, Toast.LENGTH_SHORT).show()
            }

            // TODO we need to load SSL certificates
            override fun onReceivedSslError(view: WebView, handler: SslErrorHandler?, error: SslError?) {
                if (!certPermissionsShown) {
                    var message = context.getString(R.string.dialog_message_ssl_generic)
                    when (error?.primaryError) {
                        SslError.SSL_UNTRUSTED -> message = context.getString(R.string.dialog_message_ssl_untrusted)
                        SslError.SSL_EXPIRED -> message = context.getString(R.string.dialog_message_ssl_expired)
                        SslError.SSL_IDMISMATCH -> message = context.getString(R.string.dialog_message_ssl_mismatch)
                        SslError.SSL_NOTYETVALID -> message = context.getString(R.string.dialog_message_ssl_not_yet_valid)
                    }
                    message += context.getString(R.string.dialog_message_ssl_continue)
                    AlertDialog.Builder(context, R.style.CustomAlertDialog)
                            .setTitle(context.getString(R.string.dialog_title_ssl_error))
                            .setMessage(message)
                            .setPositiveButton(android.R.string.ok, DialogInterface.OnClickListener { _, _ ->
                                certPermissionsShown = true
                                handler?.proceed()
                            })
                            .setNegativeButton(android.R.string.cancel, DialogInterface.OnClickListener { _, _ ->
                                certPermissionsShown = false
                                handler?.cancel()
                            })
                            .show()
                } else {
                    // we have already shown permissions, no need to show again on page refreshes or when page auto-refreshes itself
                    handler?.proceed()
                }
            }
        }
        screenSaverWebView?.loadUrl(url)
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun configureWebSettings(userAgent: String) {
        val webSettings = screenSaverWebView?.settings
        webSettings?.javaScriptEnabled = true
        webSettings?.domStorageEnabled = true
        webSettings?.databaseEnabled = true
        webSettings?.setAppCacheEnabled(true)
        webSettings?.javaScriptCanOpenWindowsAutomatically = true
        if (!TextUtils.isEmpty(userAgent)) {
            webSettings?.userAgentString = userAgent
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webSettings?.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }
        Timber.d(webSettings?.userAgentString)
    }

    private fun clearCache() {
        screenSaverWebView?.clearCache(true)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            CookieManager.getInstance().removeAllCookies(null)
        }
    }

    companion object {
        const val PRECIP_AMOUNT: Double = 0.3 // rain probability
        const val UNSPLASH_IT_URL = "http://unsplash.it/%s/%s?random"
    }
}